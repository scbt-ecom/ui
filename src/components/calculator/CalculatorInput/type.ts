export type TErrorInput = string | undefined;
