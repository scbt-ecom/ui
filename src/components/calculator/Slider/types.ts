import type { SliderProps as MuiSliderProps } from "@mui/material";

type SliderBaseProps = Pick<MuiSliderProps, "className">;

export interface CustomSliderProps extends SliderBaseProps, MuiSliderProps {
  defaultValue: number;
  min: number;
  max: number;
  circle?: number;
}

export interface ITestSlider {
  left: number;
  width: number;
  height: number;
  toJSON: () => void;
  x: number;
  y: number;
  bottom: number;
  right: number;
  top: number;
}
