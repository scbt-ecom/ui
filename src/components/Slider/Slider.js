import React, { useState } from 'react';
import Slider from '@material-ui/core/Slider';
import TextField from './TextField';
import NumberFormat from 'react-number-format'; //https://github.com/s-yadav/react-number-format

import useStyles from './styles';

const SliderComponent = React.memo(props => {
  const classes = useStyles(props);
  const [value, setValue] = useState(
    props.initialValue || props.value || props.min
  );

  const handleSliderChange = (_event, newValue) => {
    if (value === newValue) {
      return;
    }
    setValue(newValue);
    props.onChange(newValue);
  };

  const handleSliderChangeCommitted = (_event, newValue) => {
    if (props.onChangeCommitted) {
      setValue(newValue);
      props.onChangeCommitted(newValue);
    }
  };

  const handleInputChange = ({ floatValue = 0 }) => {
    if (floatValue !== value) {
      setValue(floatValue);
      if (props.onChangeCommitted) {
        return props.onChangeCommitted(floatValue);
      }
      props.onChange(floatValue);
    }
  };

  const handleInputBlur = e => {
    const { min, max } = props;
    let newValue = 0;

    if (value <= min) {
      newValue = min;
    } else if (value >= max) {
      newValue = max;
    }

    if (newValue) {
      setValue(newValue);
      if (props.onChangeCommitted) {
        return props.onChangeCommitted(newValue);
      }
      props.onChange(newValue);
    }
  };
  const { sliderProps, inputProps, label, min, max } = props;

  return (
    <div className={classes.container}>
      <NumberFormat
        {...inputProps}
        label={label}
        className={classes.input}
        customInput={TextField}
        thousandSeparator={' '}
        value={value}
        onValueChange={handleInputChange}
        fullWidth
        allowNegative={false}
        decimalScale={0}
        onBlur={handleInputBlur}
        InputProps={{
          readOnly: props.discrete,
        }}
      />
      <Slider
        {...sliderProps}
        classes={{
          root: classes.sliderRoot,
          markLabel: classes.sliderMarkLabel,
          mark: classes.sliderMark,
        }}
        value={value}
        onChange={handleSliderChange}
        onChangeCommitted={handleSliderChangeCommitted}
        min={min}
        max={max}
      />
    </div>
  );
});

SliderComponent.defaultProps = {
  onChange: () => null,
};

export default SliderComponent;
