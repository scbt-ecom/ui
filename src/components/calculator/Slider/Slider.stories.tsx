import type { Meta, StoryObj } from "@storybook/react";
import { Slider } from ".";

const meta: Meta<typeof Slider> = {
  title: "Calculator/Slider",
  component: Slider,
  tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Standard: Story = {
  args: {
    defaultValue: 5,
    max: 10,
    min: 0,
  },
};
