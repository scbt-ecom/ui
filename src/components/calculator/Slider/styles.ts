import { styled } from "@mui/material/styles";
import { Slider as MuiSlider } from "@mui/material";
import type { CustomSliderProps } from "./types";

export const CustomSlider = styled(MuiSlider)<CustomSliderProps>(({ circle = "32" }) => ({
  "& .MuiSlider-markLabel": {
    display: "none",
  },
  "& .MuiSlider-track": {
    borderRadius: "initial",
  },
  "& .MuiSlider-thumb": {
    width: `${circle}px`,
    height: `${circle}px`,
    border: "3.5px solid #003791",
    color: "#fff",
    "&:focus, &:hover, &.Mui-active": {
      boxShadow: "0 0 0 8px rgba(0,0,0,.1)",
    },
  },
  "& .css-1gv0vcd-MuiSlider-track": {
    border: "none",
  },
  "& .css-14pt78w-MuiSlider-rail": {
    color: "#f2f2f2",
  },
}));
