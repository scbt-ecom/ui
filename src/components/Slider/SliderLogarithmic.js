/**
 * BASED ON https://noteskeeper.ru/607/
 */

import React, { useState } from 'react';
import Slider from '@material-ui/core/Slider';
import TextField from '../TextField';
import NumberFormat from 'react-number-format'; //https://github.com/s-yadav/react-number-format
import { fromSlider, toSlider, round } from './utils';

import useStyles from './styles';

const SliderComponent = React.memo(props => {
  const classes = useStyles(props);
  const [value, setValue] = useState(
    props.initialValue || props.value || props.min
  );

  const handleSliderChange = (event, newValue) => {
    const viewValue = round(fromSlider(newValue));
    if (value === viewValue) {
      return;
    }
    setValue(viewValue);
    props.onChange(viewValue);
  };

  const handleSliderChangeCommitted = (event, newValue) => {
    const viewValue = round(fromSlider(newValue));
    if (props.onChangeCommitted) {
      setValue(viewValue);
      props.onChangeCommitted(viewValue);
    }
  };

  const handleInputChange = ({ floatValue = 0 }) => {
    if (floatValue !== value) {
      setValue(floatValue);
      if (props.onChangeCommitted) {
        return props.onChangeCommitted(floatValue);
      }
      props.onChange(floatValue);
    }
  };

  const handleInputBlur = e => {
    const { min, max } = props;
    let newValue = 0;

    if (value <= min) {
      newValue = min;
    } else if (value >= max) {
      newValue = max;
    }

    if (newValue) {
      setValue(newValue);
      if (props.onChangeCommitted) {
        return props.onChangeCommitted(newValue);
      }
      props.onChange(newValue);
    }
  };
  const { sliderProps, inputProps, label, min, max } = props;

  return (
    <div className={classes.container}>
      <NumberFormat
        {...inputProps}
        label={label}
        className={classes.input}
        customInput={TextField}
        thousandSeparator={' '}
        value={value}
        onValueChange={handleInputChange}
        fullWidth
        allowNegative={false}
        decimalScale={0}
        onBlur={handleInputBlur}
      />
      <Slider
        {...sliderProps}
        defaultValue={value}
        value={toSlider(value)}
        onChange={handleSliderChange}
        onChangeCommitted={handleSliderChangeCommitted}
        min={toSlider(min)}
        max={toSlider(max)}
        step={0.01}
      />
    </div>
  );
});

SliderComponent.defaultProps = {
  onChange: () => null,
};

export default SliderComponent;
