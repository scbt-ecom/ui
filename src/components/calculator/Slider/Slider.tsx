import { useState } from "react";
import { CustomSlider } from "./styles";
import type { CustomSliderProps } from "./types";

export const Slider = ({ defaultValue, ...props }: CustomSliderProps) => {
  const [value, setValue] = useState(defaultValue);

  const handleSliderChange = (event: Event, newValue: number | number[]) => {
    setValue(newValue as number);
  };

  return (
    <CustomSlider
      data-testid="SliderTestId"
      defaultValue={defaultValue}
      {...props}
      value={typeof value === "number" ? value : 0}
      onChange={handleSliderChange}
    />
  );
};
