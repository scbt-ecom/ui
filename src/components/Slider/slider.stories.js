import React, { useState } from "react"
// Import the storybook libraries
import { storiesOf } from "@storybook/react"
import { action } from "@storybook/addon-actions"
import { muiTheme } from "storybook-addon-material-ui"

// Import our component from this folder
import Slider from "./Slider"
import SliderLogarithmic from "./SliderLogarithmic"

import theme from "../../style/theme"
import { inputProps, sliderProps } from "./story.config"

const SliderWrapper = props => {
  const [step, setStep] = React.useState(1000)

  const getStepValue = sliderVal => {
    if (sliderVal < 300000) {
      return 1000
    }
    if (sliderVal < 1000000) {
      return 5000
    }
    return 100000
  }

  const handleOnChangeCommitted = value => {
    props.onChangeCommitted(value)
  }
  const handleOnChange = value => {
    setStep(getStepValue(value))
  }

  const dynamicStepSliderProps = {
    step: step
  }

  return (
    <>
      <Slider
        onChange={handleOnChange}
        onChangeCommitted={handleOnChangeCommitted}
        inputProps={inputProps}
        sliderProps={dynamicStepSliderProps}
        min={150000} // if no defaultValue using min value as initial
        max={30000000}
      />
    </>
  )
}

storiesOf("Slider", module)
  .addDecorator(muiTheme([theme]))
  .add("Default", () => (
    <Slider
      onChange={action("onChange")}
      label={"Сумма кредита"}
      inputProps={inputProps}
      sliderProps={sliderProps}
      // defaultValue={300}
      min={100} // if no defaultValue using min value as initial
      max={2000}
      // value={500}
    />
  ))
  .add("Committed", () => (
    <Slider
      onChangeCommitted={action("onChangeCommitted")}
      label={"Сумма кредита"}
      inputProps={inputProps}
      sliderProps={sliderProps}
      min={150000}
      max={30000000}
    />
  ))
  .add("Dynamic step", () => <SliderWrapper onChangeCommitted={action("onChangeCommitted")} />)
  .add("Logarithmic", () => (
    <SliderLogarithmic
      // onChange={action('onChange')}
      onChangeCommitted={action("onChangeCommitted")}
      inputProps={inputProps}
      sliderProps={sliderProps}
      min={150000}
      max={30000000}
      defaultValue={2000000}
    />
  ))
  .add("Dynamic values", () => {
    const [min, setMin] = useState(100000)
    const [max, setMax] = useState(1000000)
    const [sum, SetSum] = useState(100000)

    return (
      <>
        <div
          style={{
            display: "flex",
            justifyContent: "space-around",
            margin: "20px"
          }}
        >
          <span>Sum = {sum}</span>
          <button
            onClick={() => {
              SetSum(200000)
            }}
          >
            Set sum = 200000
          </button>

          <button
            onClick={() => {
              setMin(200000)
            }}
          >
            Set min = 200000
          </button>
          <button
            onClick={() => {
              setMax(500000)
            }}
          >
            Set max = 500000
          </button>
          <button
            onClick={() => {
              SetSum(100000)
              setMin(100000)
              setMax(1000000)
            }}
          >
            reset
          </button>
        </div>
        <Slider
          // onChange={action('onChange')}
          onChangeCommitted={SetSum}
          inputProps={inputProps}
          sliderProps={sliderProps}
          min={min}
          max={max}
          value={sum}
        />
      </>
    )
  })
