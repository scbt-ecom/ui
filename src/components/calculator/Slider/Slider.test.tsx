import { render, screen, fireEvent } from "@testing-library/react";
import Slider from "@mui/material/Slider";

import type { ITestSlider } from "./types";

// Mocking getBoundingClientRect
jest.spyOn(HTMLElement.prototype, "getBoundingClientRect").mockImplementation(
  () =>
    ({
      left: 19.572917938232422,
      width: 563.5208740234375,
      toJSON: () => {},
    } as ITestSlider)
);

describe("slider Component", () => {
  it("updates value when slider thumb is moved", () => {
    expect.hasAssertions();
    render(<Slider defaultValue={5} min={0} max={10} />);
    const slider = screen.getByRole("slider");

    fireEvent.mouseDown(slider, { clientX: 200 });
    fireEvent.mouseMove(slider, { clientX: 200 });
    fireEvent.mouseUp(slider);

    expect(slider).toHaveAttribute("aria-valuenow", "3");
    expect(slider).toMatchSnapshot();
  });
});
