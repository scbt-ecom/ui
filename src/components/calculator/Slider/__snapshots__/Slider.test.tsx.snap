// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Component updates value when slider thumb is moved 1`] = `
<input
  aria-orientation="horizontal"
  aria-valuemax="10"
  aria-valuemin="0"
  aria-valuenow="3"
  data-index="0"
  max="10"
  min="0"
  step="1"
  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
  type="range"
  value="3"
/>
`;

exports[`slider Component updates value when slider thumb is moved 1`] = `
<input
  aria-orientation="horizontal"
  aria-valuemax="10"
  aria-valuemin="0"
  aria-valuenow="3"
  data-index="0"
  max="10"
  min="0"
  step="1"
  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
  type="range"
  value="3"
/>
`;

exports[`updates value when slider thumb is moved 1`] = `
<input
  aria-orientation="horizontal"
  aria-valuemax="10"
  aria-valuemin="0"
  aria-valuenow="3"
  data-index="0"
  max="10"
  min="0"
  step="1"
  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
  type="range"
  value="3"
/>
`;
